core/src/nucleus/ribosome/api/mod.rs

```diff
@@ -349,33 +347,53 @@ pub fn call(
                 zome_call.fn_name.clone().as_str(),
                 &[RuntimeValue::I32(encoded_allocation_of_input as i32)],
                 mut_runtime,
-            )?
+            )
+            .map_err(|err| HolochainError::RibosomeFailed(err.to_string()))?
             .unwrap()
             .try_into()
             .unwrap();
     }
 
     // Handle result returned by invoked function
     let maybe_allocation = decode_encoded_allocation(returned_encoded_allocation);
+    let return_text: String;
+    let return_result: HcResult<String>;
     match maybe_allocation {
         // Nothing in memory, log return code
         Err(return_code) => {
-            runtime
-                .context
-                .log(&format!(
-                    "Zome Function '{}' returned: {}",
-                    zome_call.fn_name,
-                    return_code.to_string()
-                ))
-                .expect("Logger should work");
+            return_text = return_code.to_string();
+            return_result = match return_code {
+                RibosomeReturnCode::Success => Ok(String::new()),
+                RibosomeReturnCode::Failure(err_code) => {
+                    Err(HolochainError::RibosomeFailed(err_code.to_string()))
+                }
+            };
         }
         // Something in memory, try to read it
         Ok(valid_allocation) => {
             let result = runtime.memory_manager.read(valid_allocation);
-            runtime.result = String::from_utf8(result).unwrap();
+            let maybe_zome_result = String::from_utf8(result);
+            match maybe_zome_result {
+                Err(err) => {
+                    return_text = err.to_string();
+                    return_result = Err(HolochainError::RibosomeFailed(err.to_string()));
+                }
+                Ok(json_str) => {
+                    return_text = json_str.clone();
```

Comment:
it seems like nothing at all actually happens with `return_text` am I wrong?

Reply:
Its used right after in the log call.
I guess we can thank rusfmt for making that instruction unreadable.

Reply:
oh yeah I see. Yeah I totally missed it, and had a feeling somehow that I did